helix-core/src/main/java/org/apache/helix/controller/rebalancer/AbstractRebalancer.java

```diff
@@ -352,60 +356,137 @@ public static int getStateCount(String state, StateModelDefinition stateModelDef
-        if (assigned.contains(instance) || !liveAndEnabled.contains(instance)) {
+        String peekInstance = preferredActiveInstanceQueue.peek();
+        if (assignedInstances.contains(peekInstance)) {
+          preferredActiveInstanceQueue.poll();
           continue; // continue checking for the next available instance
         }
-        String proposedInstance = instance;
-        // Additional check and alternate the assignment for reducing top state handoff.
-        if (state.equals(stateModelDef.getTopState()) && !stateModelDef.getSecondTopStates()
-            .contains(currentStateMap.getOrDefault(instance, stateModelDef.getInitialState()))) {
-          // If the desired state is the top state, but the instance cannot be transited to the
-          // top state in one hop, try to keep the top state on current host or a host with a closer
-          // state.
-          for (String currentStatePrioritizedInstance : sortedPreferenceList) {
-            if (!assigned.contains(currentStatePrioritizedInstance) && liveAndEnabled
-                .contains(currentStatePrioritizedInstance)) {
-              proposedInstance = currentStatePrioritizedInstance;
-              break;
-            }
-          }
-          // Note that if all the current top state instances are not assignable, then we fallback
-          // to the default logic that assigning the state according to preference list order.
+        String proposedInstance = adjustInstanceIfNecessary(state, peekInstance,
+            currentStateMap.getOrDefault(peekInstance, stateModelDef.getInitialState()),
+            stateModelDef, assignedInstances, totalCandidateCount - assignedInstances.size(),
+            stateCount, currentStatePrioritizedInstanceIter);
+
+        if (proposedInstance.equals(peekInstance)) {
+          // If the peeked instance is the final decision, then poll it from the queue.
+          preferredActiveInstanceQueue.poll();
         }
-        // Assign the desired state to the proposed instance
+        // else, if we found a different instance for the partition placement, then we need to
+        // check the same instance again or it will not be assigned with any partitions.
+
+        // Assign the desired state to the proposed instance if not on ERROR state.
         if (HelixDefinedState.ERROR.toString().equals(currentStateMap.get(proposedInstance))) {
           bestPossibleStateMap.put(proposedInstance, HelixDefinedState.ERROR.toString());
         } else {
           bestPossibleStateMap.put(proposedInstance, state);
           stateCount--;
         }
-        assigned.add(proposedInstance);
+        // Note that in either case, the proposed instance is considered to be assigned with a state
+        // by now.
+        if (!assignedInstances.add(proposedInstance)) {
+          throw new AssertionError(String
+              .format("The proposed instance %s has been already assigned before.",
+                  proposedInstance));
```

Comment:
This should not happen, right. You do the assignedInstances both for peek of the queue and after adjust instance.

Reply:
Yeah, this is for the sanity check. If something goes wrong, I hope it fails early. This could be very hard to debug. And if we have some other customized SM defs, it might not be covered by the unit test. Adding this assertion will help us identify issues much easier.